packages/bigquery/src/index.js

```diff
@@ -250,50 +250,51 @@ BigQuery.prototype.timestamp = function(value) {
  * @return {string} - The type detected from the value.
  */
 BigQuery.getType_ = function(value) {
-  if (value instanceof BigQuery.date) {
-    return 'DATE';
-  }
-
-  if (value instanceof BigQuery.datetime) {
-    return 'DATETIME';
-  }
-
-  if (value instanceof BigQuery.time) {
-    return 'TIME';
-  }
-
-  if (value instanceof BigQuery.timestamp) {
-    return 'TIMESTAMP';
-  }
-
-  if (value instanceof Buffer) {
-    return 'BYTES';
-  }
-
-  if (is.array(value)) {
-    return 'ARRAY';
-  }
-
-  if (is.bool(value)) {
-    return 'BOOL';
-  }
+  var typeName;
 
-  if (is.number(value)) {
-    return value % 1 === 0 ? 'INT64' : 'FLOAT64';
-  }
-
-  if (is.object(value)) {
-    return 'STRUCT';
+  if (value instanceof BigQuery.date) {
+    typeName = 'DATE';
+  } else if (value instanceof BigQuery.datetime) {
+    typeName = 'DATETIME';
+  } else if (value instanceof BigQuery.time) {
+    typeName = 'TIME';
+  } else if (value instanceof BigQuery.timestamp) {
+    typeName = 'TIMESTAMP';
+  } else if (value instanceof Buffer) {
+    typeName = 'BYTES';
+  } else if (is.array(value)) {
+    return {
+      type: 'ARRAY',
+      arrayType: BigQuery.getType_(value[0])
+    };
+  } else if (is.bool(value)) {
+    typeName = 'BOOL';
+  } else if (is.number(value)) {
+    typeName = value % 1 === 0 ? 'INT64' : 'FLOAT64';
+  } else if (is.object(value)) {
+    return {
+      type: 'STRUCT',
+      structTypes: Object.keys(value).map(function(prop) {
+        return {
+          name: prop,
+          type: BigQuery.getType_(value[prop])
+        };
+      })
+    };
+  } else if (is.string(value)) {
+    typeName = 'STRING';
   }
 
-  if (is.string(value)) {
-    return 'STRING';
+  if (!typeName) {
+    throw new Error([
+      'This value could not be translated to a BigQuery data type.',
+      value
+    ].join('\n'));
   }
 
-  throw new Error([
-    'This value could not be translated to a BigQuery data type.',
-    value
-  ].join('\n'));
+  return {
```

Comment:
We used to early return with the string value. Now that we are returning an object, there was too much repetition.